# Exploring eBay Car Sales Data

In this project, we will be looking at used car sales on *eBay Kleinanzeigen*, a classified section of the German eBay website. The dataset we will be using is a modified version from the original that was scraped and uploaded to Kaggle by user [oregsleka](https://www.kaggle.com/orgesleka). While the original dataset is no longer available on Kaggle, it can be found [here](https://data.world/data-society/used-cars-data). Our modified version has 50,000 data points out of the 370,000 original points, and it has been dirtied for the purpose of practicing data cleaning. 

The goal of this project is to clean the modified dataset we are using and analyze the used car listings. 

Here are the columns we are working with:
- `dateCrawled`: When the ad was first crawled. All field values are taken from this date.
- `name`: Name of the car.
- `seller`: Whether the seller is private or a dealer.
- `offerType`: The type of listing.
- `price`: The price on the ad to sell the car.
- `abtest`: Whether the listing is included in an A/B test.
- `vehicleType`: The vehicle Type.
- `yearOfRegistration`: The year in which the car was first registered.
- `gearbox`: The transmission type.
- `powerPS`: The power of the car in PS. 
- `model`: The car model name.
- `kilometer`: How many kilometers the car has driven.
- `monthOfRegistration`: The month in which the car was first registered.
- `fuelType`: What type of fuel the car uses.
- `brand`: The brand of the car.
- `notRepairedDamage`: If the car has a damage which is not yet repaired.
- `dateCreated`: The date on which the eBay listing was created.
- `nrOfPictures`: The number of pictures in the ad.
- `postalCode`: The postal code for the location of the vehicle.
- `lastSeenOnline`: When the crawler saw this ad last online.

In [1]:
import numpy as np
import pandas as pd
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Adjusting Column Names and Values

One thing that is immediately noticeable is the content of the data is in German (as expected). To make this data a little easier to work with, we will have to translate the content. Other things to note include: the `price` and `odometer`  columns have a dtype of object; the `vehicleType`, `gearbox`, `model`, `fuelType`, and `notRepairedDamage` columns have null values; and the column names are [camelcase](https://en.wikipedia.org/wiki/Camel_case) instead of [snakecase](https://en.wikipedia.org/wiki/Snake_case), which means we are unable to replace spaces with underscores. 

In [5]:
autos.rename(columns={'dateCrawled' : 'date_crawled',
                      'offerType' : 'offer_type',
                      'vehicleType' : 'vehicle_type',
                      'yearOfRegistration' : 'registration_year',
                      'powerPS' : 'power_ps',
                      'monthOfRegistration' : 'registration_month',
                      'fuelType' : 'fuel_type',
                    'notRepairedDamage' : 'unrepaired_damage',
                     'dateCreated': 'ad_created',
                     'nrOfPictures' : 'nr_of_pictures',
                     'postalCode' : 'postal_code',
                     'lastSeen' : 'last_seen'}, inplace = True)
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [6]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-16 21:50:53  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Columns that we can probably drop include: `seller` and `offer_type`. They mostly have one value (privat and Angebot respectively) thus we would not be able to garner any useful information from them. 

Now, let's change the `price` and `odometer` columns to numeric types, and then check for outliers.

In [7]:
autos['price'] = autos['price'].str.replace(',', '')
autos['price'] = autos['price'].str.replace('$', '').astype(float)

autos['odometer'] = autos['odometer'].str.replace(',', '')
autos['odometer'] = autos['odometer'].str.replace('km', '').astype(float)
autos.rename(columns = {'odometer' : 'odometer_km'}, inplace = True)

In [8]:
autos['price'].unique().shape

(2357,)

In [9]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

A price of `$0` seems unreasonable, and a price of `$100,000,000` also seems unreasonable.

In [10]:
autos['price'].value_counts().sort_index(ascending = True).head(20)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
Name: price, dtype: int64

It appears that unfathomably low prices are somewhat frequent. 1,421 of the listings have a price of `$0`, which comprise less than `3%` of our data points. Given that eBay is an auction site, we will consider keeping points above `$1`.

In [11]:
autos['price'].value_counts().sort_index(ascending = False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price, dtype: int64

It looks like the prices jump from `$350,000` to `$999,990+`. We'll remove the points above `$350,000` as the next price is of an unrealistic increment.

In [12]:
new_autos = autos[autos['price'].between(1, 350000)]

Let's see if anything is out of the ordinary in the `registration_year` column.

In [22]:
new_autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Strangely enough, the `registration_year` column has a minimum of 1000 and a maximum of 9999. These are both impossible considering cars did not exist in the year 1000 and the year 9999 has yet to happen. We may consider dropping these listings or changing their years to not impact any statistical analysis. A car cannot be registered after the listing was seen, and the earliest year that a car could be registered may be within the early 1900s. We will ignore listings outside the range of  1900 - 2016.

In [79]:
new_autos = new_autos[new_autos['registration_year'].between(1900, 2016)]
autos = autos[autos['registration_year'].between(1900, 2016)]
new_autos['registration_year'].value_counts(normalize = True)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
          ...   
1938    0.000021
1948    0.000021
1927    0.000021
1931    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64

## Data Exploration

In [13]:
new_autos['odometer_km'].unique().shape

(13,)

There are only 13 unique values in the `odometer_km` column. The values here must be in increments instead of exact numbers.

In [14]:
new_autos['odometer_km'].describe()

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [15]:
new_autos['odometer_km'].value_counts().sort_index(ascending = True).head()

5000.0     836
10000.0    253
20000.0    762
30000.0    780
40000.0    815
Name: odometer_km, dtype: int64

In [16]:
new_autos['odometer_km'].value_counts().sort_index(ascending = False).head()

150000.0    31414
125000.0     5057
100000.0     2115
90000.0      1734
80000.0      1415
Name: odometer_km, dtype: int64

Nothing looks out of the ordinary in the `odometer_km` column. However, it appears that a majority of the listings have a value of 150,000 km. These listings most likely have an odometer value above 150,000 but there is no telling by how much for each listing.

In [17]:
new_autos.describe(include = 'all')

date_crawled         name  seller offer_type          price  \
count                 48565        48565   48565      48565   48565.000000   
unique                46882        37470       2          1            NaN   
top     2016-03-25 19:57:10  Ford_Fiesta  privat    Angebot            NaN   
freq                      3           76   48564      48565            NaN   
mean                    NaN          NaN     NaN        NaN    5888.935591   
std                     NaN          NaN     NaN        NaN    9059.854754   
min                     NaN          NaN     NaN        NaN       1.000000   
25%                     NaN          NaN     NaN        NaN    1200.000000   
50%                     NaN          NaN     NaN        NaN    3000.000000   
75%                     NaN          NaN     NaN        NaN    7490.000000   
max                     NaN          NaN     NaN        NaN  350000.000000   

       abtest vehicle_type  registration_year  gearbox      power_ps  model  \
count   48565        43979       48565.000000    46222  48565.000000  46107   
unique      2            8                NaN        2           NaN    245   
top      test    limousine                NaN  manuell           NaN   golf   
freq    25019        12598                NaN    36102           NaN   3900   
mean      NaN          NaN        2004.755421      NaN    117.197158    NaN   
std       NaN          NaN          88.643887      NaN    200.649618    NaN   
min       NaN          NaN        1000.000000      NaN      0.000000    NaN   
25%       NaN          NaN        1999.000000      NaN     71.000000    NaN   
50%       NaN          NaN        2004.000000      NaN    107.000000    NaN   
75%       NaN          NaN        2008.000000      NaN    150.000000    NaN   
max       NaN          NaN        9999.000000      NaN  17700.000000    NaN   

          odometer_km  registration_month fuel_type       brand  \
count    48565.000000        48565.000000     44535       48565   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     29368       10336   
mean    125770.101925            5.782251       NaN         NaN   
std      39788.636804            3.685595       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count              39464                48565         48565.0  48565.000000   
unique                 2                   76             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               34775                 1887             NaN           NaN   
mean                 NaN                  NaN             0.0  50975.745207   
std                  NaN                  NaN             0.0  25746.968398   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30657.000000   
50%                  NaN                  NaN             0.0  49716.000000   
75%                  NaN                  NaN             0.0  71665.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 48565  
unique                38474  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

From the remaining listings, the average price is about `$5,888.93`, with a standard deviation of `$9,059.85`. From this we can tell that a good number of the listings have a low price with quite a few having a remarkably high price. Indeed, we can see this with the price at the 75th percentile being `$7,490`. The average odometer reading is `125,770 km`, with a standard deviation of `39,788 km`. 

It is important to note that as the `date_crawled`, `ad_created`, and `last_seen` columns are strings at the moment and thus we cannot quantitatively analyze them quite yet. We will need to convert them into a numerical representation.

In [18]:
new_autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

It appears the three are formatted similarly, which makes things easier for us. The first 10 characters contain the date in each column, and the remaining 8 (after the space) contain the time. The time is not really relevant, so we will not consider it.

In [19]:
new_autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_index(ascending = True)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The dates for `date_crawled` column range from March 5th, 2016 to April 7th, 2016.

In [20]:
new_autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False).sort_index(ascending = True)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

The dates for `ad_created` column range from June 11th, 2015 to April 7th, 2016.

In [21]:
new_autos['last_seen'].str[:10].value_counts(normalize = True, dropna = False).sort_index(ascending = True)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

It appears that almost half of the sellers were last seen in the last 3 days of the dataset. 

Next, we may want to aggregate brands to see how this impacts our analyses.

In [24]:
brand_dict = {}

top_10_brands = new_autos['brand'].value_counts(normalize = True).index[:10]

for row in top_10_brands:
    list_brand = new_autos[new_autos['brand'] == row]
    mean_price = list_brand['price'].mean()
    brand_dict[row] = int(mean_price)
    
brand_dict

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749,
 'renault': 2474,
 'peugeot': 3094,
 'fiat': 2813,
 'seat': 4397}

There appears to be a big gap in average prices among the top 10 common brands, where the price jumps from `$5,402` to `$8,332`. The most common brand has an average price of `$5,402`, with the next most common being `$8,332`. The most expensive among these are Audi, Mercedes Benz, and BMW (`$9,336`, `$8,628`, and `$8,332` respectively). The least expensive among these are Opel, Fiat, and Renault (`$2,975`, `$2,813`, and `$2,474` respectively). There might be a link in average mileage and average price we have yet to investigate.

In [25]:
bd_mileage = {}

for row in top_10_brands:
    list_brand = autos[autos['brand'] == row]
    mean_mileage = list_brand['odometer_km'].mean()
    bd_mileage[row] = int(mean_mileage)
    
bd_mileage

{'volkswagen': 128955,
 'bmw': 132521,
 'opel': 129298,
 'mercedes_benz': 130886,
 'audi': 129643,
 'ford': 124131,
 'renault': 128223,
 'peugeot': 127352,
 'fiat': 117037,
 'seat': 122061}

In [26]:
bd_series = pd.Series(brand_dict)
bdm_series = pd.Series(bd_mileage)
bd_df = pd.DataFrame(bd_series, columns = ['mean_price'])
bd_df['mean_mileage'] = bdm_series

bd_df

mean_price  mean_mileage
volkswagen           5402        128955
bmw                  8332        132521
opel                 2975        129298
mercedes_benz        8628        130886
audi                 9336        129643
ford                 3749        124131
renault              2474        128223
peugeot              3094        127352
fiat                 2813        117037
seat                 4397        122061

As we can see, there is no link between average mileage and average price; Volkswagen, Audi, and Opel share relatively close average mileages yet very dissimilar average prices.

Let's find out what the most common brand/model combinations being sold are.

In [27]:
new_autos['combo'] = new_autos['brand'] + ' ' + new_autos['model']
new_autos['combo'].value_counts()

volkswagen golf      3707
bmw 3er              2615
volkswagen polo      1609
opel corsa           1592
volkswagen passat    1349
                     ... 
bmw i3                  1
audi 200                1
rover discovery         1
ford b_max              1
rover rangerover        1
Name: combo, Length: 290, dtype: int64

The most common brand/model combinations being sold include `volkswagen golf`, `bmw 3er`, and `volkswagen polo`. Upon further investigation, we found out the 2016 Volkswagen Golf vehicles have a [low reliability rating](https://cars.usnews.com/cars-trucks/volkswagen/golf/2016/reliability). The cars came with a host of issues, and the National Highway Traffic Safety Administration issued an [alarming number of recalls](https://www.carcomplaints.com/Volkswagen/Golf/2016/recalls/). Interestingly enough, the Volkswagen Group has had 1,805 car recalls per 1,000 sold [since 1985](https://www.forbes.com/sites/jimgorzelany/2016/10/24/which-automakers-have-the-lowest-recall-rates-and-the-most-serious-recalls/?sh=9e145eb7aada), marking them to have the highest recall rate.

Let's see if Volkswagen cars make up the majority of listings with a price of 0.

In [29]:
free_cars = autos[autos['price'] == 0.0]
free_cars['brand'].value_counts().head()

volkswagen    348
opel          183
bmw           154
audi          115
ford           95
Name: brand, dtype: int64

Volkswagen cars appear to make up the majority of listings with a price of 0. Let's determine if Volkswagen cars are the majority of listings overall before concluding anything.

In [80]:
totals = autos['brand'].value_counts()
proportions = (autos['brand'].value_counts() / free_cars['brand'].value_counts()).to_frame()
totals

volkswagen        10188
bmw                5284
opel               5195
mercedes_benz      4580
audi               4149
ford               3352
renault            2274
peugeot            1418
fiat               1242
seat                873
skoda               770
mazda               727
nissan              725
citroen             669
smart               668
toyota              599
sonstige_autos      526
hyundai             473
volvo               444
mini                415
mitsubishi          391
honda               377
kia                 341
alfa_romeo          318
porsche             293
suzuki              284
chevrolet           274
chrysler            176
daihatsu            123
dacia               123
jeep                108
subaru              105
land_rover           98
saab                 77
jaguar               76
trabant              75
daewoo               72
rover                65
lancia               52
lada                 29
Name: brand, dtype: int64

Indeed, we find Volkswagen vehicles making up the majority of listings.

## Conclusion

After cleaning the data, we found the average price of listings to be `$5,888.93`, with a standard deviation of `$9,059.85`. The average odometer reading is `125,770 km`, with a standard deviation of `39,788 km`. We found the most listed brands to be Volkswagen, BMW, and Opel with average prices of `$5,402`, `$8,332`, and `$2,975` respectively. The most expensive brands were Audi, Mercedes Benz, and BMW vehicles (`$9,336`, `$8,628`, `$8,332` respectively). The least expensive brands were Opel, Fiat, and Renault (`$2,975`, `$2,813`, `$2,474`).

We found no link between average mileage and average price. The most common brand/model combinations being sold were found to be Volkswagen Golf, BMW 3 Series, and Volkswagen Polo. Upon investigating, we discovered the Volkswagen Group to have the highest recall rate of any brand, [since 1985](https://www.forbes.com/sites/jimgorzelany/2016/10/24/which-automakers-have-the-lowest-recall-rates-and-the-most-serious-recalls/?sh=9e145eb7aada).